File: packages/sdk/python/robloxforge/marketplace.py
"""Marketplace search and purchase methods for the RobloxForge SDK."""

from typing import TYPE_CHECKING, Optional, List

if TYPE_CHECKING:
    from .client import Client


class Template:
    def __init__(self, data: dict):
        self.id: str = data["id"]
        self.title: str = data["title"]
        self.description: str = data.get("description", "")
        self.category: str = data["category"]
        self.price_cents: int = data["priceCents"]
        self.thumbnail_url: Optional[str] = data.get("thumbnailUrl")
        self.downloads: int = data.get("downloads", 0)
        self.average_rating: float = data.get("averageRating", 0.0)

    @property
    def price_usd(self) -> str:
        return f"${self.price_cents / 100:.2f}"

    def __repr__(self):
        return f"<Template id={self.id} title={self.title!r} price={self.price_usd}>"


class MarketplaceClient:
    def __init__(self, client: "Client"):
        self._client = client

    def search(
        self,
        query: Optional[str] = None,
        category: Optional[str] = None,
        sort_by: str = "popular",
        page: int = 1,
        limit: int = 20,
    ) -> List[Template]:
        """
        Search templates on the marketplace.

        Args:
            query: Search text
            category: Filter by category
            sort_by: Sort order (newest, popular, rating, price_asc, price_desc)
            page: Page number
            limit: Results per page (max 50)

        Returns:
            List of Template objects
        """
        params = []
        if query:
            params.append(f"q={query}")
        if category:
            params.append(f"category={category}")
        params.append(f"sort={sort_by}")
        params.append(f"page={page}")
        params.append(f"limit={limit}")

        qs = "&".join(params)
        data = self._client.request("GET", f"/api/marketplace/search?{qs}")
        return [Template(t) for t in data.get("templates", [])]

    def get(self, template_id: str) -> Template:
        """Get a single template by ID."""
        data = self._client.request("GET", f"/api/marketplace/templates/{template_id}")
        return Template(data)

    def purchase(self, template_id: str) -> dict:
        """
        Purchase a template.

        Returns:
            dict with purchaseId and downloadUrl
        """
        return self._client.request(
            "POST", f"/api/marketplace/templates/{template_id}/purchase"
        )

File: packages/sdk/typescript/src/marketplace.ts
import { RobloxForge } from './client'

export interface MarketplaceSearchOptions {
  query?: string
  category?: 'GAME_TEMPLATE' | 'MAP_TEMPLATE' | 'UI_KIT' | 'SCRIPT' | 'ASSET' | 'SOUND'
  minPrice?: number
  maxPrice?: number
  sortBy?: 'newest' | 'popular' | 'rating' | 'price_asc' | 'price_desc'
  page?: number
  limit?: number
}

export interface Template {
  id: string
  title: string
  description: string
  category: string
  priceCents: number
  thumbnailUrl: string | null
  downloads: number
  averageRating: number
  creator: { username: string; displayName: string | null }
}

export interface MarketplaceSearchResult {
  templates: Template[]
  total: number
  page: number
  limit: number
  hasMore: boolean
}

export class MarketplaceClient {
  constructor(private rf: RobloxForge) {}

  /**
   * Search templates on the marketplace.
   */
  async search(opts: MarketplaceSearchOptions = {}): Promise<MarketplaceSearchResult> {
    const params = new URLSearchParams()
    if (opts.query) params.set('q', opts.query)
    if (opts.category) params.set('category', opts.category)
    if (opts.minPrice !== undefined) params.set('minPrice', String(opts.minPrice))
    if (opts.maxPrice !== undefined) params.set('maxPrice', String(opts.maxPrice))
    if (opts.sortBy) params.set('sort', opts.sortBy)
    if (opts.page !== undefined) params.set('page', String(opts.page))
    if (opts.limit !== undefined) params.set('limit', String(opts.limit))

    const qs = params.toString()
    const res = await this.rf.request<MarketplaceSearchResult>(
      'GET',
      `/api/marketplace/search${qs ? `?${qs}` : ''}`
    )
    if (res.error) throw new Error(res.error)
    return res.data!
  }

  /**
   * Get a single template by ID.
   */
  async get(templateId: string): Promise<Template> {
    const res = await this.rf.request<Template>('GET', `/api/marketplace/templates/${templateId}`)
    if (res.error) throw new Error(res.error)
    return res.data!
  }

  /**
   * Purchase a template (requires full or assets-only scope).
   */
  async purchase(templateId: string): Promise<{ purchaseId: string; downloadUrl: string }> {
    const res = await this.rf.request<{ purchaseId: string; downloadUrl: string }>(
      'POST',
      `/api/marketplace/templates/${templateId}/purchase`
    )
    if (res.error) throw new Error(res.error)
    return res.data!
  }
}

export function marketplacePlugin(rf: RobloxForge) {
  return new MarketplaceClient(rf)
}
